### PR TITLE
46: Must buy train, 36Jr/82: ports/fishing don't count for must buy

### LINKED ISSUE #861: Corps with only a route to a port/Paris do not have a valid route requiring a train purchase
_(note: this really only applies to Paris because the ports have other revenue centers in the way so it is impossible to run a home token to a port)_

It is the train buying step and Nord does not have a valid route requiring a train purchase, but the system is prompting me to buy a train and will not allow me to skip trains.

image

{
  "status": "active",
  "actions": [
    {
      "type": "bid",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 1,
      "company": "ACC",
      "price": 20
    },
    {
      "type": "bid",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 2,
      "company": "E-SF",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 3,
      "company": "CdH",
      "price": 70
    },
    {
      "type": "bid",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 4,
      "company": "GCB",
      "price": 110
    },
    {
      "type": "bid",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 5,
      "company": "CFL",
      "price": 160
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 6,
      "company": "CFLV",
      "price": 220
    },
    {
      "type": "par",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 7,
      "corporation": "Nord",
      "share_price": "100,0,6"
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 8,
      "shares": [
        "Nord_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 9,
      "shares": [
        "Nord_2"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 10,
      "shares": [
        "Nord_3"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 11,
      "shares": [
        "Nord_4"
      ]
    },
    {
      "type": "par",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 12,
      "corporation": "B",
      "share_price": "100,0,6"
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 13,
      "shares": [
        "B_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 14,
      "shares": [
        "B_2"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 15,
      "shares": [
        "B_3"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 16,
      "shares": [
        "B_4"
      ]
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 17
    },
    {
      "type": "pass",
      "entity": "Nord",
      "entity_type": "corporation",
      "id": 18
    }
  ],
  "id": "hs_qlzbbhaf_1593013856",
  "players": [
    {
      "name": "Player 1"
    },
    {
      "name": "Player 2"
    },
    {
      "name": "Player 3"
    }
  ],
  "title": "1836Jr30",
  "description": "",
  "max_players": "3",
  "mode": "hotseat",
  "user": {
    "id": 0,
    "name": "You"
  },
  "created_at": "2020-06-24",
  "loaded": true,
  "result": {}
}

### LINKED ISSUE #948: 1846: Shouldn’t be allowed to skip train purchase
I was able to skip purchasing a train and that’s not allowed in 1846 (section 6.68 of the GMT rulebook)

### PR DESCRIPTION
**Pull request by jenf**
1846: fixes #948
1836Jr30: fixes #861
1882:  When determining whether a company has a legal route for train ownership purposes, the company president can choose whether to count the fishing exit or not.

### CODE REVIEW

File: lib/engine/game/base.rb [hunk 1]
```diff
@@ -483,6 +484,15 @@ def sellable_turn?
         @turn > 1
       end
 
+      def route?(entity)
+        @graph.route?(entity)
+      end
+
+      def must_buy_train?(entity)
+        !entity.rusted_self && entity.trains.empty? &&
+        ((MUST_BUY_TRAIN == :route && route?(entity) == true) || MUST_BUY_TRAIN == :always)
```
Comment (tobymao) [inline lib/engine/game/base.rb:493]: route?(entity) == true... tsk tsk

Comment (tobymao) [reply lib/engine/game/base.rb:493]: why does route? return an optional?

Comment (jenf) [reply lib/engine/game/base.rb:493]: would mandatory be better?

Comment (jenf) [reply lib/engine/game/base.rb:493]: Some games (1882) have tiles which can be used on a route but do not count as a route for mandatory train buying

**Comment by tobymao** [comment]
is this ready for another look?

**Comment by jenf** [comment]
yep

File: lib/engine/graph.rb [hunk 2]
```diff
@@ -20,9 +20,10 @@ def clear
       @connected_paths.clear
       @reachable_hexes.clear
       @tokens.clear
+      @routes.clear
```
Comment (tobymao) [inline lib/engine/graph.rb:23]: this is not ideal. that will make calculation expensive for all games

File: lib/engine/graph.rb [hunk 3]
```diff
@@ -116,6 +117,24 @@ def compute(corporation)
             hexes[hex.neighbors[edge]][hex.invert(edge)] = true
           end
         end
+
+        mandatory_nodes = 0
+        optional_nodes = 0
+        local_nodes.each do |p_node, _truth|
```
Comment (tobymao) [inline lib/engine/graph.rb:123]: can you do _

File: lib/engine/part/city.rb [hunk 4]
```diff
@@ -9,8 +9,8 @@ class City < RevenueCenter
       attr_accessor :reservations
       attr_reader :slots, :tokens
 
-      def initialize(revenue, slots = 1, groups = nil, hide = false, visit_cost = nil)
-        super(revenue, groups, hide, visit_cost)
+      def initialize(revenue, slots = 1, groups = nil, hide = false, visit_cost = nil, optional = false)
```
Comment (tobymao) [inline lib/engine/part/city.rb:12]: there's too much stuff here now along with the super

can we switch this to dictionary args?

def initialize(revenue, **opts)
then the super call only needs to be super

File: lib/engine/graph.rb (hunk 2 above)
Comment (jenf) [reply lib/engine/graph.rb:23]: Without it, it wouldn't recalculate properly when you were in the optional state.

Most of the time it'd be recalculating it anyway for selecting routes.

**Comment by tobymao** [comment]
replace super(rev, **opts) with just super

remove slots.fetch

**Comment by jenf** [comment]
Can we get this in, the merge to the other branch should be simple

**Pull request merged**